controller/run.go

```diff
@@ -74,3 +99,56 @@ func RunManaged(cfg config.ShipperConfig) error {
 
 	return err
 }
+
+// runController is the main runloop for the shipper itself, and managed communication with the agent.
+func runController(ctx context.Context, agentClient client.V2) error {
+	log := logp.L()
+
+	err := agentClient.Start(ctx)
+	if err != nil {
+		return fmt.Errorf("error starting connection to client")
+	}
+
+	log.Debugf("Starting error reporter")
+	go reportErrors(ctx, agentClient)
+
+	log.Debugf("Started client, waiting")
+	handler := newClientHandler()
+
+	sigc := make(chan os.Signal, 1)
+	signal.Notify(sigc, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)
+
+	// receive the units
+	for {
+		select {
+		case gotSignal := <-sigc:
+			switch gotSignal {
+			case syscall.SIGINT, syscall.SIGTERM:
+				handler.log.Info("Received sigterm/sigint, stopping")
+			case syscall.SIGHUP:
+				handler.log.Info("Received sighup, stopping")
+			}
+			handler.grpcServer.Stop()
+			handler.outputHandler.Close()
+			return nil
+		case <-ctx.Done():
+			handler.log.Info("Got context done")
+			handler.grpcServer.Stop()
+			handler.outputHandler.Close()
+			return nil
+		case change := <-agentClient.UnitChanges():
+			// Unsure of if we want to make some of these unit change operations parallel.
+			// Because we have to juggle state between two units, there's a lot of back-and-forth logic that can lead to races. Ditto with BounceShipper().
```

Comment:
This comment is now out of date. I also don't think we need to make these operations parallel, they can be sequential for simplicity.